x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -115,47 +117,77 @@ public RoleReferenceIntersection getRoleReferenceIntersection(@Nullable Anonymou
     }
 
     public boolean canAccessResourcesOf(Subject resourceCreatorSubject) {
-        if (API_KEY.equals(getType()) && API_KEY.equals(resourceCreatorSubject.getType())) {
-            final boolean sameKeyId = getMetadata().get(AuthenticationField.API_KEY_ID_KEY)
-                .equals(resourceCreatorSubject.getMetadata().get(AuthenticationField.API_KEY_ID_KEY));
-            assert false == sameKeyId || getUser().principal().equals(resourceCreatorSubject.getUser().principal())
-                : "The same API key ID cannot be attributed to two different usernames";
-            return sameKeyId;
-        } else if ((API_KEY.equals(getType()) && false == API_KEY.equals(resourceCreatorSubject.getType()))
-            || (false == API_KEY.equals(getType()) && API_KEY.equals(resourceCreatorSubject.getType()))) {
+        if (eitherIsAnApiKey(resourceCreatorSubject)) {
+            if (bothAreApiKeys(resourceCreatorSubject)) {
+                final boolean sameKeyId = isTheSameApiKey(resourceCreatorSubject);
+                assert false == sameKeyId || getUser().principal().equals(resourceCreatorSubject.getUser().principal())
+                    : "The same API key ID cannot be attributed to two different usernames";
```

Comment:
This assertion cannot be done inside `isTheSameApiKey` because currently the remote access authentication breaks it, i.e. the same API key but user principal (QC username) is different. This is sorta related to the separate discussion we are having about what the user information we should use to populate the effective subject. It's _not_ a blocker here. Just wanted to flag it.